orchagent/dash/dashaclgroupmgr.cpp

```diff
@@ -0,0 +1,721 @@
+    }
+
+    for (const auto &tag : rule.m_src_tags)
+    {
+        auto prefixes = m_dash_acl_orch->getDashAclTagMgr().getPrefixes(tag);
+
+        src_prefixes.insert(src_prefixes.end(),
+            prefixes.begin(), prefixes.end());
+    }
+
+    for (const auto &tag : rule.m_dst_tags)
+    {
+        auto prefixes = m_dash_acl_orch->getDashAclTagMgr().getPrefixes(tag);
+
+        dst_prefixes.insert(dst_prefixes.end(),
+            prefixes.begin(), prefixes.end());
+    }
+
+    attrs.emplace_back();
+    attrs.back().id = SAI_DASH_ACL_RULE_ATTR_SRC_PORT;
+    attrs.back().value.u16rangelist.count = static_cast<uint32_t>(rule.m_src_ports.size());
+    attrs.back().value.u16rangelist.list = rule.m_src_ports.data();
+
+    attrs.emplace_back();
+    attrs.back().id = SAI_DASH_ACL_RULE_ATTR_DST_PORT;
+    attrs.back().value.u16rangelist.count = static_cast<uint32_t>(rule.m_dst_ports.size());
+    attrs.back().value.u16rangelist.list = rule.m_dst_ports.data();
+
+    if (src_prefixes.size())
+    {
+        attrs.emplace_back();
+        attrs.back().id = SAI_DASH_ACL_RULE_ATTR_SIP;
+        attrs.back().value.ipprefixlist.count = static_cast<uint32_t>(src_prefixes.size());
+        attrs.back().value.ipprefixlist.list = src_prefixes.data();
+    }
+
+    if (dst_prefixes.size())
+    {
+        attrs.emplace_back();
+        attrs.back().id = SAI_DASH_ACL_RULE_ATTR_DIP;
+        attrs.back().value.ipprefixlist.count = static_cast<uint32_t>(dst_prefixes.size());
+        attrs.back().value.ipprefixlist.list = dst_prefixes.data();
+    }
+
+    attrs.emplace_back();
+    attrs.back().id = SAI_DASH_ACL_RULE_ATTR_DASH_ACL_GROUP_ID;
+    attrs.back().value.oid = group.m_dash_acl_group_id;
+
+    auto status = sai_dash_acl_api->create_dash_acl_rule(&rule.m_dash_acl_rule_id, gSwitchId, static_cast<uint32_t>(attrs.size()), attrs.data());
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        return status;
+    }
+
+    CrmResourceType crm_rtype = (group.m_ip_version == SAI_IP_ADDR_FAMILY_IPV4) ?
+            CrmResourceType::CRM_DASH_IPV4_ACL_RULE : CrmResourceType::CRM_DASH_IPV6_ACL_RULE;
+    gCrmOrch->incCrmDashAclUsedCounter(crm_rtype, group.m_dash_acl_group_id);
+
+    return status;
+}
+
+task_process_status DashAclGroupMgr::createRule(const string& group_id, const string& rule_id, DashAclRule& rule)
+{
+    SWSS_LOG_ENTER();
+
+    auto group_it = m_groups_table.find(group_id);
+    if (group_it == m_groups_table.end())
+    {
+        SWSS_LOG_INFO("ACL group %s doesn't exist, waiting for group creating before creating rule %s", group_id.c_str(), rule_id.c_str());
+        return task_need_retry;
+    }
+    auto& group = group_it->second;
+
+    auto acl_rule_it = group.m_dash_acl_rule_table.find(rule_id);
+    ABORT_IF_NOT(acl_rule_it == group.m_dash_acl_rule_table.end(), "Failed to create ACL rule %s. Rule already exist in ACL group %s", rule_id.c_str(), group_id.c_str());
```

Comment:
createRule method should be called only if the rule doesn't exist. Otherwise, updateRule should be called. It is the responsibility of the called to check and call the right method. The assertion is to prevent coding errors. It should not happen in a normal run.